crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -143,49 +205,440 @@ impl Acir {
+                    InternalVar::default()
+                } else {
+                    todo!("dynamic arrays are not implemented yet");
+                }
+            }
         };
 
+        output_var.idx = Some(ins.idx);
+
         self.arith_cache.insert(ins.idx, output_var);
     }
+
+    pub fn print_field(f: FieldElement) -> String {
+        if f == -FieldElement::one() {
+            return "-".to_string();
+        }
+        if f == FieldElement::one() {
+            return String::new();
+        }
+        let big_f = BigUint::from_bytes_be(&f.to_bytes());
+
+        if big_f.to_string() == "4294967296" {
+            return "2^32".to_string();
+        }
+        let s = big_f.bits();
+        let big_s = BigUint::one() << s;
+        if big_s == big_f {
+            return format!("2^{}", s);
+        }
+        if big_f == BigUint::zero() {
+            return "0".to_string();
+        }
+        if big_f.clone() % BigUint::from(2_u128).pow(32) == BigUint::zero() {
+            return format!("2^32*{}", big_f / BigUint::from(2_u128).pow(32));
+        }
+        let big_minus = BigUint::from_bytes_be(&(-f).to_bytes());
+        if big_minus.to_string().len() < big_f.to_string().len() {
+            return format!("-{}", big_minus);
+        }
+        big_f.to_string()
+    }
+
+    pub fn print_gate(g: &Gate) -> String {
+        let mut result = String::new();
+        match g {
+            Gate::Arithmetic(a) => {
+                for i in &a.mul_terms {
+                    result += &format!(
+                        "{}x{}*x{} + ",
+                        Acir::print_field(i.0),
+                        i.1.witness_index(),
+                        i.2.witness_index()
+                    );
+                }
+                for i in &a.linear_combinations {
+                    result += &format!("{}x{} + ", Acir::print_field(i.0), i.1.witness_index());
+                }
+                result += &format!("{} = 0", Acir::print_field(a.q_c));
+            }
+            Gate::Range(w, s) => {
+                result = format!("x{} is {} bits", w.witness_index(), s);
+            }
+            Gate::Directive(Directive::Invert { x, result: r }) => {
+                result = format!("1/{}={}, or 0", x.witness_index(), r.witness_index());
+            }
+            _ => {
+                dbg!(&g);
+            }
+        }
+
+        result
+    }
+
+    pub fn print_circuit(gates: &[Gate]) {
+        for gate in gates {
+            println!("{}", Acir::print_gate(gate));
+        }
+    }
+
+    //Load array values into InternalVars
+    //If create_witness is true, we create witnesses for values that do not have witness
+    pub fn load_array(
+        &mut self,
+        array: &MemArray,
+        array_index: u32,
+        create_witness: bool,
+        evaluator: &mut Evaluator,
+    ) -> Vec<InternalVar> {
+        let mut result: Vec<InternalVar> = Vec::new();
+        for i in 0..array.len {
```

Comment:
How is this different? would the code be more efficient?

Reply:
I am not a big fan of it, I think it is more clear to use a for loop and a push rather than using a mapping, a closure and the collect but I guess this is more rusty..!